h/settings.py

```diff
@@ -13,54 +14,90 @@ class SettingError(Exception):
     pass
 
 
-class DeprecatedSetting(object):
-    """A wrapper for deprecated settings which emits appropriate warnings."""
-
-    def __init__(self, setting, message):
-        self.setting = setting
-        self.message = message
-
-        # Test seam
-        self.warn = log.warn
-
-    def __call__(self, environ):
-        result = self.setting(environ)
-        if result is not None:
-            self.warn(self.warning)
-        return result
-
-    @property
-    def warning(self):
-        return 'use of {s} is deprecated: {m}'.format(s=self.setting,
-                                                      m=self.message)
-
-
-class EnvSetting(object):
-
-    """An (optionally typed) setting read from an environment variable."""
-
-    def __init__(self, setting, varname, type=None):
-        self.setting = setting
-        self.varname = varname
-        if type is not None:
-            self.type = type
-        else:
-            self.type = str
-
-    def __call__(self, environ):
-        if self.varname in environ:
+class SettingsManager(object):
+    """
+    Configuration setting resolver.
+
+    SettingsManager resolves settings from various sources into the final typed
+    values used when the app runs. It also provides a way to check for missing
+    required settings or use of deprecated settings.
+
+    The resolved settings are available via the `settings` attribute.
+    """
+
+    def __init__(self, settings={}, environ=None):
```

Comment:
> There's a bug here - the default value for settings should be None, not {}, due to the notorious Python mutable default argument behaviour. 

What is the bug? The default value is never exposed as a reference (it is merely used in the call to `self.settings.update(settings)` so it cannot be modified.

Reply:
Hmm.. I think you're right! Because `__init__()` never modifies the default `settings` dict then, _as long as `settings` never contains any mutable sub-objects_, there's technically nothing broken here. I was wondering why none of the tests seemed to be broken.

I think I agree with pylint that mutable default args should just be avoided in Python though - it would make it very easy for later changes to this class or I think even for some code that uses a list or dict as a setting value to introduce a really subtle bug into the tests. So glad to see you've removed this.